Add regex search

fourshade opened the pull request:
I have made a few changes to the program mainly for my personal use, but I feel like it's time to share them now. There is now a "Regex Search" checkbox on the lookup window that allows regular expression search of the loaded dictionaries rather than just exact match w/ prefixes and suffixes:

![image](https://user-images.githubusercontent.com/39717492/43353434-f73fa1d4-91fe-11e8-945e-be4ff126686c.png)

Even without using regex symbols, this search can handle partial word searches, which has been a requested feature. The main thing I was worried about was the computational power required to filter() 80,000 dictionary keys using a regular expression. The performance doesn't seem too bad on my computer, but mine is rather fast and may not be a good benchmark. It's also possible that Python has built-in optimizations for this sort of thing, which would be great, but I just don't know for sure. If it does perform well, it could eventually be a good replacement for the normal search (with regex special characters escaped).

Also, if you enter a raw steno stroke pattern in the lookup window (i.e. STROEBG), it will return what it maps to. Not a huge change, but it does help to identify strokes that produce something other than readable text.

benoit-pierre commented:
Thanks, but this should really be split in to 3 separate pull requests: update get-pip SHA1, add regex support to the dictionary lookup dialog, and another one for the new config engine command.

I've fixed get-pip SHA1 on master, please rebase with only the changes to add support for the regex functionality.

fourshade commented:
Okay, I've reworked my fork's history and split up the changes.

benoit-pierre commented on .gitignore:
```diff
@@ -40,9 +40,13 @@ nosetests.xml
+# Vim swap/undo files
+*.swp
+*.un
+
```
This does not belong in the PR.

benoit-pierre commented on plover/steno_dictionary.py:
```diff
@@ -245,6 +251,15 @@ def casereverse_lookup(self, value):
+            mkpairs = ((m,self.reverse_lookup(m)) for m in sorted(matches))
+            return [(m,k) for (m,k) in mkpairs if k][:match_limit]
+        except re.error as e:
+            return [("Invalid regular expression", [(e.msg,)])]
```
That error handling does not look right: exception handling should be done in the caller.

benoit-pierre commented on plover/steno_dictionary.py:
```diff
@@ -264,7 +279,7 @@ def save(self, path_list=None):
         If <path_list> is None, all writable dictionaries are saved'''
         if path_list is None:
-            dict_list = [d for d in self if dictionary.save is not None]
+            dict_list = [d for d in self if not d.readonly]
```
Why?

fourshade replied:
It checks for the existence of a save method rather than the readonly property. Even if a dictionary has this method we may not want it saved if readonly is set.

(also my IDE threw up an error over the name "dictionary", I don't see it defined in that namespace.)

fourshade commented on plover/gui_qt/suggestions_widget.py:
```diff
@@ -52,6 +52,12 @@ def append(self, suggestion_list):
+        self.clear()
+        self.append(suggestion_list)
+        scrollbar = self.suggestions.verticalScrollBar()
+        scrollbar.setValue(scrollbar.minimum())
```
That change was made because, if the search returns enough results to overfill the suggestions window, it will automatically scroll down. This is desirable if you're using the suggestions widget to show suggestions as you type, but in the lookup window it's not, because usually the first results are the ones you want to see. Only the lookup code calls this.

fourshade commented on plover/gui_qt/lookup_dialog.py:
```diff
@@ -36,10 +38,21 @@ def eventFilter(self, watched, event):
+                suggestion_list = self._engine.get_regex_suggestions(translation)
+            else:
+                suggestion_list = self._engine.get_suggestions(translation)
+            if matchStenoString.match(translation):
```
I'm getting rid of that one. Too much change for too little benefit. Was actually thinking about a checkbox for it.

fourshade commented on plover/steno_dictionary.py:
```diff
@@ -159,7 +161,11 @@ def reverse_lookup(self, value):
         return list(self.casereverse[value].keys())
-
+    
+    def regex_value_match(self, rpattern, match_limit=100):
```
That particular limit is just there as sort of a sanity check in case the user searches for something insane like .*
I tried to set a maximum that would be more than enough to populate the lookup window in any given case. Can be moved into a constant.

fourshade commented on plover/steno_dictionary.py:
```diff
@@ -245,6 +251,15 @@ def casereverse_lookup(self, value):
             if key:
                 return key
 
+    def regex_reverse_lookup(self, pattern, match_limit=10):
```
That one I'd like to move into a config value, something like "max_search_results". I'm working on a computationally cheaper partial word lookup as well; it could work for both limits.

fourshade commented on plover/suggestions.py:
```diff
@@ -50,3 +50,16 @@ def find(self, translation):
 
         return suggestions
+    
+    def findRegex(self, translation):
```
Fixed those. Got an IDE that that checks for PEP 8 now.

fourshade commented:
I've cleaned up all of this stuff, and in the process I've folded the regex stuff into a larger overhaul of the lookup system that uses bisection of a sorted keys list to find words that begin with the search string (i.e. if you type "sitt", it will return the word "sitting" along with others that might be close even if there is no explicit entry that maps to "sitt"). Iteration over tens of thousands of strings with regex is probably too expensive for a search that runs for every character you type, but it's still handy enough at times that I left it there as a checkbox, off by default.

Since the sort is case-insensitive, it also replaces the need for a separate case-reverse lookup dict and saves about 10 MB of memory (the case-reverse dictionary was a dict of dicts with a ton of overhead).  I added a lot more comments this time, but it is still a relatively major change, and I still need to benchmark and write some good tests for it before I push to the fork. Should I just close this pull request and open a new one?

fourshade commented:
Well I have a couple new ideas, and more has changed, so I'm gonna close this PR and maybe open others with more incremental changes.
